zfs/zfs.go

```diff
@@ -1032,6 +1037,34 @@ type RecvOptions struct {
 	RollbackAndForceRecv bool
 	// Set -s flag used for resumable send & recv
 	SavePartialRecvState bool
+	InheritProperties    []Property
+	OverrideProperties   map[Property]string
+}
+
+type Property string
+
+func (p Property) Validate() error {
+	// see zfsprops(8), section "User Properties"
+	if len(p) < 1 {
+		return fmt.Errorf("property name can't be empty")
+	}
+	if len(p) > 256 {
+		return fmt.Errorf("property name longer than 256 characters")
+	}
+	if p[0] == '-' {
+		return fmt.Errorf("property name can't start with '-'")
+	}
+	propertyValidChar := regexp.MustCompile(`^[0-9a-zA-Z-_\.:]+$`)
+	if !propertyValidChar.MatchString(string(p)) {
+		return fmt.Errorf("property name must only contain alphanumeric chars and any in %q", "-_.:")
```

Comment:
you can `.String()` a regex

Reply:
This is an error message i copy-paste stole from you :) I don't think "doesn't match regex: [ascii puke]" is help- or meaningful to most users. The current message is somewhat more 'outspoken' than `^[0-9a-zA-Z-_\.:]+$`.

Reply:
agreed, but then either put comments at the regex definition and fmt.Errof sites, or move the fmt.Errorf closer to the `propertyValidNameChars` definition like so:

```
// Keep regex and error message in sync!
var (
  propertyValidNameChars = regexp.MustCompile(`^[0-9a-zA-Z-_\.:]+$`)
  propertyValidNameCharsErr = fmt.Errorf("property name must only contain alphanumeric chars and any in %q", "-_.:")
)
```